patches/api/0332-Better-Stats-API.patch

```diff
@@ -0,0 +1,721 @@
++    public static final CustomStatistic POT_FLOWER = create("pot_flower");
++    public static final CustomStatistic TRIGGER_TRAPPED_CHEST = create("trigger_trapped_chest");
++    public static final CustomStatistic OPEN_ENDERCHEST = create("open_enderchest");
++    public static final CustomStatistic ENCHANT_ITEM = create("enchant_item");
++    public static final CustomStatistic PLAY_RECORD = create("play_record");
++    public static final CustomStatistic INTERACT_WITH_FURNACE = create("interact_with_furnace");
++    public static final CustomStatistic INTERACT_WITH_CRAFTING_TABLE = create("interact_with_crafting_table");
++    public static final CustomStatistic OPEN_CHEST = create("open_chest");
++    public static final CustomStatistic SLEEP_IN_BED = create("sleep_in_bed");
++    public static final CustomStatistic OPEN_SHULKER_BOX = create("open_shulker_box");
++    public static final CustomStatistic OPEN_BARREL = create("open_barrel");
++    public static final CustomStatistic INTERACT_WITH_BLAST_FURNACE = create("interact_with_blast_furnace");
++    public static final CustomStatistic INTERACT_WITH_SMOKER = create("interact_with_smoker");
++    public static final CustomStatistic INTERACT_WITH_LECTERN = create("interact_with_lectern");
++    public static final CustomStatistic INTERACT_WITH_CAMPFIRE = create("interact_with_campfire");
++    public static final CustomStatistic INTERACT_WITH_CARTOGRAPHY_TABLE = create("interact_with_cartography_table");
++    public static final CustomStatistic INTERACT_WITH_LOOM = create("interact_with_loom");
++    public static final CustomStatistic INTERACT_WITH_STONECUTTER = create("interact_with_stonecutter");
++    public static final CustomStatistic BELL_RING = create("bell_ring");
++    public static final CustomStatistic RAID_TRIGGER = create("raid_trigger");
++    public static final CustomStatistic RAID_WIN = create("raid_win");
++    public static final CustomStatistic INTERACT_WITH_ANVIL = create("interact_with_anvil");
++    public static final CustomStatistic INTERACT_WITH_GRINDSTONE = create("interact_with_grindstone");
++    public static final CustomStatistic TARGET_HIT = create("target_hit");
++    public static final CustomStatistic INTERACT_WITH_SMITHING_TABLE = create("interact_with_smithing_table");
++
++    /**
++     * Get a custom stat type from a key
++     *
++     * @param key the stats key
++     * @return the stat or null if none found
++     */
++    @Nullable
++    public static CustomStatistic getByKey(@NotNull NamespacedKey key) {
++        return CUSTOM_STATISTIC_MAP.get(key);
++    }
++
++    /**
++     * Gets an immutable collection of all custom stat types.
++     *
++     * @return a collection of all custom stat types
++     */
++    @NotNull
++    public static Collection<CustomStatistic> values() {
++        return CUSTOM_STATISTIC_MAP.values();
++    }
++
++    private final NamespacedKey key;
++    private final boolean isUsed;
```

Comment:
The isUsed solution looks a bit inelegant - removed stat types should be deprecated and not added, to not forever exist as unnecessary noise in iteration